lib/seattleflu/id3c/cli/command/etl/redcap.py

```diff
@@ -44,3 +69,630 @@ def normalize_net_id(net_id: str=None) -> Optional[str]:
+        birth_date: str = None, zipcode: str = None, unique_identifier: str = None) -> tuple:
+    """
+    Returns a FHIR Patient resource entry and reference.
+    Uses demographics to create the patient identifier unless
+    a *unique_identifier* is provided.
+    """
+
+    if not unique_identifier and (not first_name or not last_name or not birth_date \
+        or not zipcode):
+        LOG.debug('If you are not providing a `unique_identifier` you should provide' \
+            +' `first_name`, `last_name`, `birth_date`, and `zipcode`')
+
+    gender = map_sex(sex)
+
+    language_codeable_concept = create_codeable_concept(
+        system = 'urn:ietf:bcp:47',
+        code = preferred_language
+    )
+
+    communication = [{
+        'language' : language_codeable_concept,
+        'preferred': True
+    }]
+
+    patient_id = None
+    if unique_identifier:
+        patient_id = generate_hash(unique_identifier)
+    else:
+        patient_id = generate_patient_hash(
+                names       = (first_name, last_name),
+                gender      = gender,
+                birth_date  = birth_date,
+                postal_code = zipcode)
+
+    if not patient_id:
+        # Some piece of information was missing, so we couldn't generate a
+        # hash.  Fallback to treating this individual as always unique by using
+        # the REDCap record id.
+        patient_id = generate_hash(f"{record.project.base_url}{record.project.id}/{record.id}")
+
+    LOG.debug(f"Generated individual identifier {patient_id}")
+
+    patient_identifier = create_identifier(f"{system_identifier}/individual", patient_id)
+    patient_resource = create_patient_resource([patient_identifier], gender, communication)
+
+    return create_entry_and_reference(patient_resource, "Patient")
+
+
+def create_patient_using_demographics(sex: str, preferred_language: str, first_name: str, last_name: str,
+        birth_date: str, zipcode: str, record: REDCapRecord, system_identifier: str) -> tuple:
+    """
+    Returns a FHIR Patient resource entry and reference.
+    Uses demographics to create the patient identifier
+    """
+    return _create_patient(
+        sex = sex,
+        preferred_language = preferred_language,
+        first_name = first_name,
+        last_name = last_name,
+        birth_date = birth_date,
+        zipcode = zipcode,
+        record = record,
+        system_identifier = system_identifier)
+
+
+def create_patient_using_unique_identifier(sex: str, preferred_language: str, unique_identifier: str,
+    record: REDCapRecord, system_identifier: str) -> tuple:
+    """
+    Returns a FHIR Patient resource entry and reference.
+    Uses a unique identifier to create the patient identifier
+    """
+    return _create_patient(
+        sex = sex,
+        preferred_language = preferred_language,
+        record = record,
+        system_identifier = system_identifier,
+        unique_identifier = unique_identifier)
```

Comment:
I want to continue the conversation that I had started in the [previous PR](https://github.com/seattleflu/id3c-customizations/pull/152#discussion_r506711947). This is definitely _not_ a blocker, I just want to understand the reasoning of using two overloads vs using the inner function with different arguments. (Would love to learn if this is a coding practice that I'm not aware of).

Reply:
The optional parameters on  _create_patient are: 
1 first_name
2 last_name
3 birth_date
4 zipcode
5 unique_identifier

But they are *conditionally* optional.

If you provide a value for 5, you don't need to provide a value for 1 through 4. But if you don't provide a value for 5, you should provide a value for 1, 2, 3, and 4. (I've added a debug statement to warn if any of the values of 1 through 4 are None when 5 is None. I'm sure there will be cases when we don't have 4 but have 1, 2, and 3, so I didn't want to error on that.)

So that the caller doesn't have to keep track of which parameters are required in which scenarios I made the two overloads in which all parameters are required.

Reply:
Ah, I see. Thank you for the explanation 🙏